### DISCUSSION TITLE
UNIPELT, UNIPELT (AP) and UNIPELT (APL)

### DISCUSSION THREAD

**Discussion by keyuchen21** [adapter-hub/adapters#696] [Q&A] [answered]
1. I'm wondering what's the different between UNIPELT implementation in Adapters and UNIPELT (AP)/UNIPELT (APL) in the original paper([Mao et al. (2022)](https://arxiv.org/pdf/2110.07577.pdf))?

2. are there any AP and APL version of UNIPELT in Adapters already?
3. is AP and APL version just two/three UNIPELT stack on top of each others?

![38451713395269_ pic](https://github.com/adapter-hub/adapters/assets/54015474/3fbb9bc6-2c38-47a7-981b-56a93c3be34c)

**Comment by calpt** [comment, marked as answer]
Hey, the UniPELT AP and APL variants in the paper compose different types of adapters. Specifically:

**UniPELT APL** combines bottleneck adapters (A), prefix tuning (P) and LoRA (L). This is the default version in _Adapters_ if you use `UniPELTConfig()`:
```python
config = UniPELTConfig()
model.add_adapter("unipelt_apl", config=config)
```
To make this clearer, we can define the same config via `ConfigUnion()`:
```python
config = ConfigUnion(
    PrefixTuningConfig(prefix_length=10, use_gating=True),
    SeqBnConfig(reduction_factor=16, use_gating=True),
    LoRAConfig(r=8, alpha=2, use_gating=True)
)
model.add_adapter("unipelt_apl", config=config)
```
This is identical to using `UniPELTConfig()`

**UniPELT AP** combines bottleneck adapters (A), prefix tuning (P). We can also define this using `ConfigUnion()` in _Adapters_:
```python
config = ConfigUnion(
    PrefixTuningConfig(prefix_length=10, use_gating=True),
    SeqBnConfig(reduction_factor=16, use_gating=True)
)
model.add_adapter("unipelt_ap", config=config)
```

Hope this helps!

**Reply by keyuchen21** [reply]
thank you a lot !!!!!!!!!!!!!!!!!!!!!!!
